charon/tests/ui/arrays.out

```diff
@@ -1655,6 +1655,109 @@ fn test_crate::sum_mut_slice<'_0>(@1: &'_0 mut (Slice<u32>)) -> u32
     return
 }
 
+fn test_crate::slice_pattern_1(@1: Array<(), 1 : usize>)
+{
+    let @0: (); // return
+    let x@1: Array<(), 1 : usize>; // arg #1
+    let _named@2: (); // local
+    let @3: (); // anonymous local
+    let @4: &'_ (Array<(), 1 : usize>); // anonymous local
+    let @5: &'_ (()); // anonymous local
+
+    @fake_read(x@1)
+    @4 := &x@1
+    @5 := @ArrayIndexShared<'_, (), 1 : usize>(move (@4), const (0 : usize))
+    _named@2 := copy (*(@5))
+    @3 := ()
+    @0 := move (@3)
+    drop _named@2
+    @0 := ()
+    return
+}
+
+fn test_crate::slice_pattern_2<'_0, T>(@1: Array<&'_0 mut (T), 3 : usize>)
+{
+    let @0: (); // return
+    let x@1: Array<&'_ mut (T), 3 : usize>; // arg #1
+    let _a@2: &'_ mut (T); // local
+    let _b@3: &'_ mut (T); // local
+    let _c@4: &'_ mut (T); // local
+    let @5: (); // anonymous local
+    let @6: &'_ mut (Array<&'_ mut (T), 3 : usize>); // anonymous local
+    let @7: &'_ mut (&'_ mut (T)); // anonymous local
+    let @8: &'_ mut (Array<&'_ mut (T), 3 : usize>); // anonymous local
+    let @9: &'_ mut (&'_ mut (T)); // anonymous local
+    let @10: &'_ mut (Array<&'_ mut (T), 3 : usize>); // anonymous local
+    let @11: &'_ mut (&'_ mut (T)); // anonymous local
+
+    @fake_read(x@1)
+    @10 := &mut x@1
+    @11 := @ArrayIndexMut<'_, &'_ mut (T), 3 : usize>(move (@10), const (0 : usize))
```

Comment:
We have a `simplify_constants` pass that moves constant operands into separate locals but unfortunately it runs before `index_to_function_calls`. I would want it to apply here, so we don't get a `const` as function argument. Unfortunately `simplify_constants` operates on ullbc and `index_to_function_calls` operates on llbc so we can't easily reorder them... We might have to rewrite `simplify_constants` to operate on llbc, it's not the first time I've wanted that.

Reply:
We can do this later, I'll ask Son about it

Reply:
Son says it's fine

Reply:
Clarifying here: `simplify_constants` it transforms the AST to remove some cases of constants altogether (for instance: constant adts) so that they don't duplicate other operand cases, and it does so by moving some constants to auxiliary operands (among other things).